x/leverage/keeper/oracle.go

```diff
@@ -24,26 +26,65 @@ func (k Keeper) TokenPrice(ctx sdk.Context, denom string) (sdk.Dec, error) {
 		return sdk.ZeroDec(), types.ErrBlacklisted
 	}
 
-	price, err := k.oracleKeeper.GetExchangeRateBase(ctx, denom)
+	price, err := k.oracleKeeper.GetExchangeRateBase(ctx, baseDenom)
 	if err != nil {
 		return sdk.ZeroDec(), sdkerrors.Wrap(err, "oracle")
 	}
 
 	if price.IsNil() || !price.IsPositive() {
-		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrInvalidOraclePrice, denom)
+		return sdk.ZeroDec(), sdkerrors.Wrap(types.ErrInvalidOraclePrice, baseDenom)
 	}
 
 	return price, nil
 }
 
+// TokenSymbolPrice returns the USD value of a token. Note, the input denom must
+// still be the base denomination, e.g. uumee. When error is nil, price is guaranteed
+// to be positive. Also returns the token's exponent to reduce redundant registry reads.
+func (k Keeper) TokenSymbolPrice(ctx sdk.Context, baseDenom string) (sdk.Dec, uint32, error) {
+	t, err := k.GetTokenSettings(ctx, baseDenom)
+	if err != nil {
+		return sdk.ZeroDec(), 0, err
+	}
+
+	if t.Blacklist {
+		return sdk.ZeroDec(), t.Exponent, types.ErrBlacklisted
+	}
+
+	price, err := k.oracleKeeper.GetExchangeRate(ctx, t.SymbolDenom)
+	if err != nil {
+		return sdk.ZeroDec(), t.Exponent, sdkerrors.Wrap(err, "oracle")
+	}
+
+	if price.IsNil() || !price.IsPositive() {
+		return sdk.ZeroDec(), t.Exponent, sdkerrors.Wrap(types.ErrInvalidOraclePrice, baseDenom)
```

Comment:
ok, we have zero check here :+1:

Reply:
Yeah, all negative & zero & nil checks are handled when initially fetching prices from oracle, since there are only two (previous one) such points in the code.

Everything else acts on the assumption of a positive (or error) return.